atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/AbortedTransactionSoftCache.java

```diff
@@ -0,0 +1,166 @@
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+public final class AbortedTransactionSoftCache implements AutoCloseable {
+    public enum TransactionSoftCacheStatus {
+        PENDING_LOAD_FROM_RELIABLE,
+        IS_ABORTED,
+        IS_NOT_ABORTED;
+    }
+
+    private final DisruptorAutobatcher<Long, TransactionSoftCacheStatus> autobatcher;
+
+    private volatile PatchyCache patchyCache = null;
+    private final FutileTimestampStore futileTimestampStore;
+    private final KnownConcludedTransactions knownConcludedTransactions;
+
+    public AbortedTransactionSoftCache(
+            FutileTimestampStore futileTimestampStore, KnownConcludedTransactions knownConcludedTransactions) {
+        this.futileTimestampStore = futileTimestampStore;
+        this.knownConcludedTransactions = knownConcludedTransactions;
+        this.autobatcher = Autobatchers.coalescing(this::processBatch)
+                .safeLoggablePurpose("get-transaction-soft-cache-status")
+                .batchFunctionTimeout(Duration.ofSeconds(30))
+                .build();
+    }
+
+    public TransactionSoftCacheStatus getSoftCacheTransactionStatus(long startTimestamp) {
+        return AtlasFutures.getUnchecked(autobatcher.apply(startTimestamp));
+    }
+
+    private Map<Long, TransactionSoftCacheStatus> processBatch(Set<Long> request) {
+        Optional<PatchyCache> maybeSnapshot = getSnapshot();
+
+        long latestTsSeenSoFar = getLatestTsSeenSoFar(request, maybeSnapshot);
+        long latestBucketSeenSoFar = Utils.getBucket(latestTsSeenSoFar);
+
+        PatchyCache refreshedPatchyCache = refreshPatchyCache(maybeSnapshot, latestTsSeenSoFar, latestBucketSeenSoFar);
+
+        return KeyedStream.of(request)
+                .map(startTimestamp -> {
+                    long requestedBucket = Utils.getBucket(startTimestamp);
+
+                    if (requestedBucket < latestBucketSeenSoFar) {
+                        return TransactionSoftCacheStatus.PENDING_LOAD_FROM_RELIABLE;
+                    } else {
+                        return getStatus(startTimestamp, refreshedPatchyCache.abortedTransactions);
+                    }
+                })
+                .collectToMap();
+    }
+
+    private long getLatestTsSeenSoFar(Set<Long> batch, Optional<PatchyCache> maybeSnapshot) {
+        long latestTsSeenSoFar = batch.stream().max(Comparator.naturalOrder()).orElse(0L);
+
+        if (maybeSnapshot.isPresent()) {
+            latestTsSeenSoFar = Math.max(latestTsSeenSoFar, maybeSnapshot.get().lastKnownConcludedTimestamp);
+        }
+
+        long latestBucket = Utils.getBucket(latestTsSeenSoFar);
+        // The purpose of this call is to refresh the knownConcluded store for current bucket if it is not up-to-date.
+        // Do not remove this line without considering perf implications.
+        knownConcludedTransactions.isKnownConcluded(latestBucket, KnownConcludedTransactions.Consistency.REMOTE_READ);
```

Comment:
This call is made with a timestamp (and not a bucket) if I'm not mistaken, and skips the DB call if `latestBucket`, which is a small number, is cached. Do we need the max timestamp in the bucket?

Reply:
great catch! you are correct, i think this became lost during a refactor for a previous comment. I will add some tests for this,
I had considered stronger type casting for this reason but seemed like too much on this hot path. Curious what you think

Reply:
discussed offline: I think this is not super super hot given what the other classes on this path already do, generally speaking (though it's a good thing to be cautious!)